### PR TITLE
feat: For Shipping Zones Zip Regex - Correctly handle UK postcodes redacted by Apple Pay

### LINKED ISSUE #2645: Apple Pay postcode backfill blocks address checks
**Describe the bug**
When the shipping address is supplied by Apple Pay, it is only initially a partial postcode, backfilled in this plugin with `*` symbols. This backfilling makes it impossible to do certain address checks correctly for shipping types.

In our case, shipping from London to central Brighton (BN1 postcode) has one pricing structure, while shipping further west such as Lancing (BN15) has a different price.

But since the postcode is padded with `*` characters, the same as the wildcard character, we cannot differentiate between `BN1*` and `BN15*` with Apple Pay.

**To Reproduce**
1. Have separate shipping restrictions for `BN1 *` vs `BN15*` - note that we cannot just insert a rule for `BN1*` because that will match BN1 and BN15 when it shouldn't.
2. Add a shipping address in Apple Pay for a BN1 postcode.
3. Attempt to buy - the shipping address is rejected since `BN1****` doesn't match either `BN1 *` or any other `BN1x` (e.g. `BN15*`) rules

**Expected behavior**

**Screenshots**

**Environment (please complete the following information):**
 - WordPress Version - 6.2.2
 - WooCommerce Version - 7.7.0
 - Stripe Plugin Version - 7.4.1
 - Browser [e.g. chrome, safari] and Version - iOS 16.5
 - Any other plugins installed - not relevant

**Additional context**

The change to fix this is very simple in class-wc-stripe-payment-request.php, in method get_normalized_postal_code, the following code is present:

```
		if ( 'GB' === $country ) {
			// Replaces a redacted string with something like LN10***.
			return str_pad( preg_replace( '/\s+/', '', $postcode ), 7, '*' );
		}
```

This just needs to use any character other than * to be able to differentiate between a digit and a wildcard - I've tested this with # and it seems to work as I'd expect.

### PR DESCRIPTION
**Pull request by majikandy**
Improved upon idea in #2646 So that user does not need to add special syntax into shipping zones regex

Fixes #2645 

## Changes proposed in this Pull Request:

UK Postcodes consist of an outcode 2,3 or 4 characters long, then an optional space, and then 3 more characters. 

Apple Pay intelligently truncates the UK postcode removing the outcode. Futhermore It keeps the optional space at the end of it's redacted string if it was there in the original.

With this behaviour known, we need to code it more intelligently when padding with stars than just assuming a UK postcode is 7 characters long.

This PR implements that behaviour.

This also enabled full support for the Regex within Shipping Zones so that Regex like

`N1 *` (N1 space star) correctly match postcodes like N12AA and N1 2AA to mean all N1 postcodes and correctly avoids accidental matches of N10, N11, N12 postcodes etc

This enabled proper bevaviour of shipping zones such that a user can ship to N1 for one price and N19 for another. The main scenario of this is an artistan seller/maker who will drive or walk to N1 to hand deliver, but won't to N19 without a fee (or even not at all)

# Questions for the PR author:
- How can this code break?
  If Apple changed the way it redacts the postcodes
- What are we doing to make sure this code doesn't break?
if the truncated postcode excluding space is more than 5 characters, it just returns it as is as it is highly likely it is has not been truncated. This is because it would be a bad implementation by a third party to truncate a postcode to 5 chars for privacy reasons, when many true valid postcodes are 5 characters long.

Also the only impact of the postcode not validating is that shipping zones come up wrong due to non matches of the truncated/altered postcode. Therefore the behaviour cannot be made any worse than now.

## Testing instructions

- Admin: Enable Apple pay via Stripe on the Basket
- Admin: Setup shipping zones with a regex like 'N1 *' for free and 'N10 * for a fee.

- Customer: Setup addresses in your real apple pay on your device  eg N1 4AB, N10 5BC, N11AA, N191AA (you can set any addresses as you don't need to complete purchases or validate addresses)
- Customer: Add items to the basket
- Customer: for each of the addresses in apple pay
   - Trigger apple pay, 
   - choose address, 
   - see correct option. 
   - Now press cancel, refresh the page, 
   - see the redacted address like N1 *** near the delivery options on the screen (also here the options should be correctly available as it uses the same code to determine it and the triggering of apple pay actually updates that address)

Unit tests not added as this piece of code yet as it didn't have any before - however can easily add some if this change is likely to progress into the codebase

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 1]
```diff
@@ -328,6 +328,27 @@ public function get_normalized_postal_code( $postcode, $country ) {
...
+			// 'LN10 1AA' |  'LN10 '  |  'LN10 ***'
+			// 'LN101AA'  |  'LN10'   |  'LN10 ***'
+			// 'W10 2AA'  |  'W10 '   |  'W10 ***'
+			// 'W102AA'   |  'W10'    |  'W10 ***'
+			// 'N2 3AA    |  'N2 '    |  'N2 ***'
+			// 'N23AA     |  'N2'     |  'N2 ***'
+
+			$spaceless_postcode = preg_replace( '/\s+/', '', $postcode );
+
+			if ( strlen($spaceless_postcode) < 5) {
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php]:
```suggestion
			if ( strlen( $spaceless_postcode ) < 5) { 
```

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 2]
```diff
@@ -328,6 +328,27 @@ public function get_normalized_postal_code( $postcode, $country ) {
...
+			// 'W10 2AA'  |  'W10 '   |  'W10 ***'
+			// 'W102AA'   |  'W10'    |  'W10 ***'
+			// 'N2 3AA    |  'N2 '    |  'N2 ***'
+			// 'N23AA     |  'N2'     |  'N2 ***'
+
+			$spaceless_postcode = preg_replace( '/\s+/', '', $postcode );
+
+			if ( strlen($spaceless_postcode) < 5) { 
+				// Always reintroduce the space so that Shipping Zones regex like 'N1 *' work to match N1 postcodes like N1 1AA, but don't match N10 postcodes like N10 1AA
+				return $spaceless_postcode . " ***";
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php]:
```suggestion
				return $spaceless_postcode . ' ***'; 
```

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 3]
```diff
@@ -512,13 +512,34 @@ public function filter_gateway_title( $title, $id ) {
...
+			// 'LN10 1AA' |  'LN10 '  |  'LN10 ***'
+			// 'LN101AA'  |  'LN10'   |  'LN10 ***'
+			// 'W10 2AA'  |  'W10 '   |  'W10 ***'
+			// 'W102AA'   |  'W10'    |  'W10 ***'
+			// 'N2 3AA    |  'N2 '    |  'N2 ***'
+			// 'N23AA     |  'N2'     |  'N2 ***'
+			
+			$spaceless_postcode = preg_replace( '/\s+/', '', $postcode );
+			
+			if ( strlen($spaceless_postcode) < 5) {
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-payment-request.php]:
```suggestion
			if ( strlen( $spaceless_postcode ) < 5) { 
```

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 4]
```diff
@@ -512,13 +512,34 @@ public function filter_gateway_title( $title, $id ) {
...
+			// 'W10 2AA'  |  'W10 '   |  'W10 ***'
+			// 'W102AA'   |  'W10'    |  'W10 ***'
+			// 'N2 3AA    |  'N2 '    |  'N2 ***'
+			// 'N23AA     |  'N2'     |  'N2 ***'
+			
+			$spaceless_postcode = preg_replace( '/\s+/', '', $postcode );
+			
+			if ( strlen($spaceless_postcode) < 5) { 
+				// Always reintroduce the space so that Shipping Zones regex like 'N1 *' work to match N1 postcodes like N1 1AA, but don't match N10 postcodes like N10 1AA
+				return $spaceless_postcode . " ***";
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-payment-request.php]:
```suggestion
				return $spaceless_postcode . ' ***'; 
```

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php (hunk 1 above)
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php]: I think we can move this hardcoded number (5) to a constant.

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 5]
```diff
@@ -328,6 +328,27 @@ public function get_normalized_postal_code( $postcode, $country ) {
 		if ( 'GB' === $country ) {
 			// Replaces a redacted string with something like LN10***.
 			return str_pad( preg_replace( '/\s+/', '', $postcode ), 7, '*' );
```
Comment (annemirasol) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php]:
```suggestion
```

I think we meant to remove the early return here.

**Comment by annemirasol** [comment]
Setup: `N1 *` Free, `N10 *` and `N11 *` 5.00
Screenshot 2024-12-20 at 11 47 26 AM


1. Testing PRB Apple Pay

| N15 (not defined) | N11 | N10 | N1 |
|-----|-----|-----|-----|
|Screenshot 2024-12-20 at 11 28 41 AM|Screenshot 2024-12-20 at 11 30 27 AM|Screenshot 2024-12-20 at 11 30 34 AM|Screenshot 2024-12-20 at 11 30 42 AM|

2. Testing ECE Apple Pay, **_after removing the early return at the top of the GB logic (see comment above)_**

| N15 (not defined) | N11 | N10 | N1 |
|-----|-----|-----|-----|
|Screenshot 2024-12-20 at 11 28 41 AM|Screenshot 2024-12-20 at 11 30 27 AM|Screenshot 2024-12-20 at 11 30 34 AM|Screenshot 2024-12-20 at 11 30 42 AM|

3. Modified the Free shipping zone regex to `N1*` (without the space):
PRB and ECE: all addresses qualified for free shipping

4. Tested Google Pay `N1C 4AG` address, with `N1 *` Free:
Screenshot 2024-12-20 at 12 02 36 PM

5. Tested Google Pay `N1C 4AG` address, with `N1*` Free:
Screenshot 2024-12-20 at 11 59 25 AM


---
Thanks for working on this! 🚢 pending suggested changes

**Review by wjrosa — APPROVED** [review]

**Comment by wjrosa** [comment]
Hey @majikandy ! Thank you for your contribution! I took the liberty of fixing and updating your PR with the suggestions and linting issues. I also added the changelog and readme entries. As well as specific unit tests. Since we tested and approved it, I decided to merge it already 👍

**Comment by majikandy** [comment]
@wjrosa Amazing, thank you. I was already using my fork effectively on my own shop so I'm really happy you (all) picked this up and got it in properly. I would have got round to it but have been taking a break from the computer for a couple of weeks. Hope you liked the change.

**Comment by wjrosa** [comment]
> @wjrosa Amazing, thank you. I was already using my fork effectively on my own shop so I'm really happy you (all) picked this up and got it in properly. I would have got round to it but have been taking a break from the computer for a couple of weeks. Hope you liked the change.

Thank you for your hard work. And have fun on your break!

**Pull request merged**